Array swap improvements

e-kayrakli opened the pull request:
This PR improves the array swap operation in terms of error checking and
applicability of pointer-swap-based swap operation.

Motivated by an issue found by @bradcray here:
  https://github.com/chapel-lang/chapel/pull/16246#discussion_r510645106

Summary of changes:
- Moves the suitability check for the optimization into its own separate helper
  in `BlockArr` implementation.

  This new check disallows optimized swaps for arrays with domains:

  ```chapel
  var d1 = {0..9} dmapped Block({1..10});
  var d2 = {1..10} dmapped Block({1..10});
  ```
  which was previously allowed. While still allowing the optimization for:

  ```chapel
  var d1 = {0..9} dmapped Block({1..10});
  var d2 = {0..9} dmapped Block({1..10});
  ```
- Limits the optimized swap to same-type arrays only.

  Arguably there can be optimizations that we can do for cross-type swaps. But
  that's future work.

- Adds a `config param debugOptimizedSwap = false`, and some output controlled
  by that flag.

- While there, adds rank and size mismatch checks for all `<=>` calls. The size
  checks are subject to `boundsChecking` being on.

- Adds bunch of more testing.

Test:
- [x] gasnet.quickstart asan on `test/optimizations/arraySwap`
- [x] gasnet.quickstart asan on `test/distributions/robust/arithmetic/basics/test_array_swap.chpl`
- [x] full standard
- [x] full gasnet

ronawho commented on modules/dists/BlockDist.chpl:
```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
+  var domsMatch = true;
+  var domShapesMatch = true;
+
+  if this.dom != other.dom { // no need to check if this is true
```
For the dynamic fast followers we just check `dom.dist.dsiEqualDMaps`. Why do we need to check the domain too in this case?

bradcray replied:
Because a domain 0..99 and one 1..100 over the same distribution aren't guaranteed to have equal-sized chunks on all locales that support pointer-swapping.  So today, we make a swap that should be legal but slow illegal.

ronawho replied:
Why isn't that a problem for fast-followers too?

bradcray replied:
I don't know, maybe it is?  You're saying fast followers don't currently check that the local number of elements to be iterated over are the same for both arrays?

e-kayrakli replied:
Fast follower checks are delegated to distributions. And at the very least Block checks whether domains' `whole`s are the same.

There is also a TODO there to relax that rule, which is already relaxed for swaps with this implementation. At least to some extent. (Arrays can be swapped fast, if the domains don't match exactly but their shapes match as long as their distribution has the some bounding box as the domain itself)

https://github.com/chapel-lang/chapel/blob/master/modules/dists/BlockDist.chpl#L1154

bradcray replied:
Just to follow on, I verified that fast followers don't fire for the same kind of 0..99 / 1..100 case mentioned above (not that I doubted Engin's analysis so much as because it wouldn't completely surprise me if we didn't have tests for it, and Elliot so rarely asks questions that don't point to an actual problem).

bradcray commented on modules/dists/BlockDist.chpl:
```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
+
+  if this.dom != other.dom { // no need to check if this is true
+    if domsMatch {
+      for param i in 0..this.dom.rank-1 {
```
This doesn't need to be a `param`, right?  Is your thought that in the typical case, avoiding the loop will give an execution-time benefit that makes the increase in code size worthwhile?  (or even that the code won't increase much in size?)

e-kayrakli replied:
Yeah you're right. It is my wishful thinking that having it `param` or not doesn't show any actual difference in either compilation or execution time. 

The main reason I have semi-automatically made this loop `param` is that the Block distribution code has several loops over ranks and all of which are `param` loops. (At least the ones I can search and find).

If you feel non-`param` is safer, I can change it for sure.

bradcray replied:
I don't know that I have a preference, I was curious as much as anything.  I think that if it were on a more obviously performance critical path (like local dsiAccess) I'd definitely use a param loop.  Whereas for a distributed memory swap, it's probably not a big win (but also may not be a big generated code size hit.  This one isn't obvious to me which it should be.  If we hadn't been talking about premature optimization of generated code size recently, I might not have even noticed it.

e-kayrakli replied:
I see. Consistency in coding style outweighs my fear of growing the AST size because of the unrolling that happens here. As you don't have strong opinions, I am leaning towards keeping this loop a `param` for now.

bradcray commented on modules/dists/BlockDist.chpl:
```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
+        if this.dom.whole.dim(i) != other.dom.whole.dim(i) {
+          if this.dom.whole.dim(i).size != other.dom.whole.dim(i).size {
+            domShapesMatch = false;
+          }
```
I'd probably be inclined to remove this `domShapesMatch=false` conditional and the subsequent "shapes match and they match their bounding boxes" conditional because it feels like a very specific special case that's unlikely to happen often in practice (i.e. I don't think typical code patterns result in this pattern).  So it feels like a bit of runtime overhead and generated code + needing to maintain / ensure correctness of this code without any obvious benefit.

A check I could imagine being more useful would be to see whether each locale involved in the swap owned a chunk of the same size, but that has greater execution time cost (so I don't think we should do it either, at least without a compelling use case).

e-kayrakli replied:
I don't feel strongly about it and can remove that check, but my thinking was that this and the subsequent check are executed only if corresponding domain dimensions don't match exactly. This implies that without these checks we'll run a `forall` to do the swap and the cost of these checks will be negligible compared to that, and that comparison only makes sense if these checks fail, too. Because otherwise, we'll get the benefit of pointer swapping.

It is common to have domains match their distributions' bounding boxes but you are suggesting that swaps between such arrays are not common, right?:

```chapel
var D1 = {1..10} dmapped Block({1..10});
var D2 = {0..9} dmapped Block({0..9});
var A1: [D1] int;
var A2: [D2] int;
A1 <=> A2;
```

In any case, left to my own devices, I'd keep the check for the reason above and that it adds coverage for a case that might come up. But I don't feel strongly about it as I see your points, too. I'll leave the final judgement call to you if you still think covering these cases doesn't worth the additional overhead on the unoptimized case.

bradcray replied:
> but you are suggesting that swaps between such arrays are not common, right?:

Right.  To me the common cases are "we're all using the same domain or distribution" or "we're each using our own domain and/or distribution, but all using the same indexing scheme", but not "we've each got our own domain and distribution and indexing scheme" (while also wanting to swap arrays and expecting it to be fast).

I agree that the execution time checks are going to be negligible.  To me, the overhead is the additional code complexity in the swap operator that future developers would need to understand when modifying/updating this code (or, in the moment, that I need to feel confident in as a reviewer).

Given that mix, I'd still be inclined to remove it, as I'd prefer to avoid carrying around too much code optimizing for cases that don't have a use case that we'd consider compelling / likely / something that anyone should naturally expect to be fast.  Though if you disagree with that characterization (i.e., you think it is more likely than I expect or something that a user should naturally expect to be fast), I'm definitely open to counterarguments.

e-kayrakli replied:
All sounds reasonable to me and I buy the argument that maintenance overhead is important here. I don't have any concrete counterargument at the moment, I'll remove this check.

bradcray commented on modules/dists/BlockDist.chpl:
```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
+
+// <=> implementation uses Reflection to check whether this resolves, so
+// it is OK for this to not resolve without debugOptimizedSwap.
+proc BlockArr.doiOptimizedSwap(other) where debugOptimizedSwap {
```
I must be misunderstanding something:  This looks as though if `debugOptimizedSwap` is true, we will never do an optimized swap (because this routine returns `false`).

e-kayrakli replied:
If `debugOptimizedSwap` is true, this overload will be called for cases where we are swapping between different types. See that the other main overload of this function has `other: this.type` as an argument.

The only reason for this overload is to provide some debugging output without actually making the main overload a generic one, thinking that fully generic functions are additional load on resolution. But that may be a premature optimization and can cause more maintenance overhead than its worth. What do you think?

bradcray replied:
I see now, thanks for the explanation.  It's a little thing, but if we keep this, I'd move it after the "actual" doiOptimizedSwap() so that someone searching for the routine (from the top of the file) or reading the code from top to bottom will come across it first rather than this overload.  I'd probably also update the comment to more proactively state why it's here, e.g., "The purpose of this overload is to provide debugging output in the event that `debugOptimizedSwap` is on and the main routine doesn't resolve (e.g., due to a type, stridability, or rank mismatch in the `other` argument).  When `debugOptimizedSwap` is off, this overload will be ignored due to its `where` clause."

My (potentially incorrect) sense is that `where` clauses also add resolution load (moreso than formal argument checks I think Michael has said), but I couldn't say how a generic function whose body returns false compares to an overload with a false where clause.  In any case, I don't have an objection to the routine or its where clause if the comment had made its role a bit clearer (the current comment focuses more on mechanism and doesn't relate it to the "real" routine at all).

e-kayrakli replied:
I'll keep the function in with the adjustments you suggest.

My thinking was that if we have a fully generic version of the function that returns false for some types, that'd mean we instantiate that function for many different type combinations only to return false and/or give a debug message. With this kind of implementation we create a doiOptimizedSwap only if there is a type match and what we need to do is some runtime checks. (Unless you explicitly set `debugOptimizedSwap`)

Do any of these cause noticeable impact for swap? probably not given it is not an integral part of the interface :)

bradcray commented on modules/dists/BlockDist.chpl:
```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
-  if(this.dom.dist.dsiEqualDMaps(other.dom.dist)) {
+proc BlockArr.doiOptimizedSwap(other: this.type) 
+    where this.stridable == other.stridable &&
+          this.rank == other.rank {
```
Is this where clause redudant with the `this.type` constraint?  My intuition says "yes", but since it's here, I worry that I'm missing something.  If so, I think it should be removed, both for simplicity, to avoid spreading superstitious coding if people start worrying that the type constraint isn't sufficient, and because of the aforementioned cost of evaluating where clauses in the compiler.

So then the "resolution performance" question for me becomes "is choosing between two overloads cheaper or more expensive than having a single generic function that we instantiate for each type we try to resolve it with?" (where again, I'm unsure... but I mention it now that I see it was generic before, and because I think the logic / choices would be easier to follow with a single routine containing control flow vs. multiple overloads that rely upon formal type constraints and where clauses to understand.  i.e., if compilation time was not an issue or we knew that it was no better, I'd definitely put everything back into a single routine again for code clarity).  I'm not at all insistent on this, just commenting to see what you think.

e-kayrakli replied:
The where clause should go away, I believe. 

But your question is interesting. My intuition as I mentioned above is that multiple overloads is better as you wouldn't instantiate a big function only to fold some param ifs later on. That being said, I don't trust in my knowledge on the order of operations during function resolution. Though it makes sense that first you need to instantiate the function than fold ifs in instantiations?

e-kayrakli replied:
There is a bug (quirk?) with canResolveMethod that these where clauses should remain or I should change `this.type` in the formal declaration to something more explicit.

```chapel
use Reflection;

record R {
  var x;

  proc foo(r: this.type) {
    compilerWarning("Resolving foo. this.type: ", this.type:string, 
                                  " r.type: ", r.type:string);
  }

  proc bar(r: R(int(64))) {
    compilerWarning("Resolving foo. this.type: ", this.type:string, 
                                  " r.type: ", r.type:string);
  }

  proc baz(r: R(this.x.type)) {
    compilerWarning("Resolving foo. this.type: ", this.type:string, 
                                  " r.type: ", r.type:string);
  }
}

var r1 = new R(5);
var r2 = new R(5.2);

compilerWarning("Can resolve foo? ", canResolveMethod(r1, "foo", r2):string);
compilerWarning("Can resolve bar? ", canResolveMethod(r1, "bar", r2):string);
compilerWarning("Can resolve baz? ", canResolveMethod(r1, "baz", r2):string);
```

Generates:

```
$CHPL_HOME/canResolveBug.chpl:6: In method 'foo':
$CHPL_HOME/canResolveBug.chpl:7: warning: Resolving foo. this.type: R(int(64)) r.type: R(real(64))
  $CHPL_HOME/modules/standard/Reflection.chpl:290: called as R(int(64)).foo(r: R(real(64))) from function 'canResolveMethod'
  $CHPL_HOME/canResolveBug.chpl:1: called as canResolveMethod(obj: _any, param fname = "foo": string, args(0): _any)
note: generic instantiations are underlined in the above callstack
$CHPL_HOME/canResolveBug.chpl:1: In module 'canResolveBug':
$CHPL_HOME/canResolveBug.chpl:25: note: Can resolve foo? true
$CHPL_HOME/canResolveBug.chpl:26: warning: Can resolve bar? false
$CHPL_HOME/canResolveBug.chpl:27: warning: Can resolve baz? false
```

Where if the formal type is `this.type` we don't match against the generic fields and try to resolve the function. This causes resolution errors in this branch. I'm recording this here to investigate further soon.

e-kayrakli replied:
OTOH, this made me realize that we should probably prevent the optimized path from executing if the arrays element types are different. As it can change the type mismatch error messages in a bad way. Last commits make that change

e-kayrakli replied:
I recorded the reflection problem in https://github.com/chapel-lang/chapel/issues/16673

bradcray replied:
OK, thanks!

bradcray commented on modules/internal/ChapelArray.chpl:
```diff
@@ -4405,6 +4409,12 @@ module ChapelArray {
+      compilerError("rank mismatch in array swap");
+
+    if boundsChecking then
+      checkArrayShapesUponAssignment(x, y, forSwap=true);
```
Thanks for noticing and fixing this case!

If for any reason you preferred, I think it would be fine not to specialize the halt for swapping vs. assigning in the sense that I think it'd be reasonable to consider a swap a form of assignment (i.e., I wouldn't expect a user to be confused if their `A <=> B` statement talked in terms of assignment.

e-kayrakli replied:
Yeah I agree that that wouldn't be a big deal. But this is more accurate and can be more helpful I think.

bradcray commented on modules/internal/DefaultRectangular.chpl:
```diff
@@ -2264,10 +2264,17 @@ module DefaultRectangular {
+  proc DefaultRectangularArr.doiOptimizedSwap(other) where debugOptimizedSwap {
+    writeln("DefaultRectangularArr doing unoptimized swap. Type mismatch");
+    return false;
+  }
```
Similar comments here to the BlockArr overloads, obviously...  Whatever we do, we should do for both similarly.

e-kayrakli commented:
@bradcray -- except for the one topic that still looks unresolved, I think I have addressed your feedback. And there has been some changes for that topic but not quite in the direction that I was imagining we can.

I am rerunning gasnet on this branch and PR message may need some changes before I merge.

I am planning to merge this early next week barring any issues with testing or any outstanding concerns by then.

bradcray commented:
I think this is generally looking OK to me.  I gave the latest version only a cursory look after the previous reviews.  Let me know if there's anything specific you'd want me to focus in on in more detail before merging.

e-kayrakli commented:
Thanks! I don't have any particular concerns for now.

Another gasnet testing came clean. I'll merge this next week.
